File: data_structures/linked_lists/doubly_linked_lists/doubly_linked_lists.c
/**
 * @author Yannick Brenning
 * @date 03.01.2021
 * @brief Implementation of doubly linked lists
 */

#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include "doubly_linked_lists.h"

node_t *dllist_init(const int val) {
    node_t *head = (node_t *) malloc(sizeof(node_t));
    assert(head);

    head->val = val;
    head->prev = NULL;
    head->next = NULL;

    return head;
}

void dllist_destroy(node_t **head) {
    if (dllist_isempty(*head)) {
        return;
    }

    node_t *curr = *head;
    node_t *temp;
    while (curr != NULL) {
        temp = curr;
        curr = curr->next;
        free(temp);
    }

    *head = NULL;
}

bool dllist_isempty(node_t *head) {
    return (head == NULL);
}

int dllist_get(node_t *head, const unsigned int index) {
    if (dllist_isempty(head)) {
        return -1;
    }

    node_t *curr = head;
    for (int i = 0; i < index; i++) {
        if (curr->next == NULL) {
            return -1;
        }

        curr = curr->next;
    }

    return curr->val;
}

void dllist_append(node_t **head, const int val) {
    if (dllist_isempty(*head)) {
        *head = dllist_init(val);
        return;
    }

    node_t *curr = *head;
    while (curr->next != NULL) {
        curr = curr->next;
    }

    node_t *new_node = (node_t *) malloc(sizeof(node_t));
    new_node->val = val;
    new_node->next = NULL;
    new_node->prev = curr;

    curr->next = new_node;
}

void dllist_push(node_t **head, const int val) {
    node_t *new_node = (node_t *) malloc(sizeof(node_t));
    new_node->val = val;
    new_node->prev = NULL;
    new_node->next = *head;

    (*head)->prev = new_node;
    *head = new_node;
}

bool dllist_insert(node_t **head, const int val, const unsigned int index) {
    if (dllist_isempty(*head) && index != 0) {
        return false;
    }

    if (index == 0) {
        dllist_push(head, val);
        return true;
    }

    node_t *curr = *head;
    for (int i = 0; i < index - 1; i++) {
        if (curr->next == NULL) {
            return false;
        }

        curr = curr->next;
    }

    node_t *new_node = (node_t *) malloc(sizeof(node_t));
    new_node->val = val;
    new_node->prev = curr;
    new_node->next = curr->next;

    curr->next = new_node;

    return true;
}

int dllist_remove_last(node_t **head) {
    int retval = -1;

    if (dllist_isempty(*head)) {
        return retval;
    }

    if ((*head)->next == NULL) {
        retval = (*head)->val;

        free(*head);
        *head = NULL;

        return retval;
    }

    node_t *curr = *head;
    while (curr->next->next != NULL) {
        curr = curr->next;
    }

    retval = curr->next->val;
    free(curr->next);
    curr->next = NULL;

    return retval;
}

int dllist_remove_first(node_t **head) {
    int retval = -1;

    if (dllist_isempty(*head)) {
        return retval;
    }

    retval = (*head)->val;

    if ((*head)->next == NULL) {

        free(*head);
        *head = NULL;

        return retval;
    }

    node_t *temp = (*head)->next;
    temp->prev = NULL;
    free(*head);
    (*head) = temp;

    return retval;
}

int dllist_delete(node_t **head, const unsigned int index) {
    int retval = -1;
    if (dllist_isempty(*head)) {
        return retval;
    }

    if (index == 0) {
        return dllist_remove_first(head);
    }

    // Get to the node before the one we want to delete
    // such that curr->next is the node to be removed
    node_t *curr = *head;
    for (int i = 0; i < index - 1; i++) {
        // Check if we are out of bounds
        if (curr->next == NULL) {
            return retval;
        }

        curr = curr->next;
    }

    if (curr->next == NULL) {
        return retval;
    }

    retval = curr->next->val;
    node_t *new_next = curr->next->next;

    if (new_next == NULL) {
        return dllist_remove_last(head);
    } else {
        new_next->prev = curr;
    }

    free(curr->next);
    curr->next = new_next;

    return retval;
}

void dllist_reverse(node_t **head) {
    node_t *curr = *head;
    node_t *temp = NULL;
    while (curr != NULL) {
        // Reversing doubly linked list simply
        // requires us to swap each prev and next node
        temp = curr->prev;
        curr->prev = curr->next;
        curr->next = temp;

        curr = curr->prev;
    }

    // Set new head to last node of old llist
    // NULL <- old_head -> <- node_1 -> ... <- temp -> <- old_tail -> NULL
    if (temp != NULL) {
        *head = temp->prev;
    }
}

void dllist_print(node_t *head) {
    if (dllist_isempty(head))
        printf("[ ]\n");
    else {
        printf("[ ");

        node_t *curr = head;
        while (curr->next != NULL) {
            printf("%d ", curr->val);
            curr = curr->next;
        }

        printf("%d ]\n", curr->val);
    }
}

File: data_structures/linked_lists/doubly_linked_lists/doubly_linked_lists.py
"""
Implementation of Doubly Linked Lists.
"""

from __future__ import annotations
from typing import Any


class Node:
    def __init__(self, val: Any) -> None:
        self.val = val
        self.next: Node | None = None
        self.prev: Node | None = None


class DoublyLinkedList:
    def __init__(self) -> None:
        self.head: Node | None = None
        self.size = 0

    def __str__(self) -> str:
        output = "[ "

        curr = self.head
        for _ in range(0, self.size):
            output += str(curr.val) + " "
            curr = curr.next

        output += "]"

        return output

    def __len__(self) -> int:
        return self.size

    def is_empty(self) -> bool:
        return self.size == 0

    def append(self, val: Any) -> None:
        """
        Appends node to end of Doubly Linked List
        :param val: value of node to append
        :return: None
        >>> linked_list = DoublyLinkedList()
        >>> linked_list.append(1)
        >>> linked_list.append(2)
        >>> linked_list.append(3)
        >>> print(linked_list)
        [ 1 2 3 ]
        """

        if self.is_empty():
            self.head = Node(val)
        else:
            curr = self.head
            while curr.next is not None:
                curr = curr.next

            new_node = Node(val)
            new_node.prev = curr

            curr.next = new_node

        self.size += 1

    def push(self, val: Any) -> None:
        """
        Push node to front of Doubly Linked List
        :param val: value of node to be added
        :return: None
        >>> linked_list = DoublyLinkedList()
        >>> linked_list.push(1)
        >>> linked_list.push(2)
        >>> linked_list.push(3)
        >>> print(linked_list)
        [ 3 2 1 ]
        """

        if self.is_empty():
            self.head = Node(val)
        else:
            new_node = Node(val)
            new_node.prev = None
            new_node.next = self.head

            self.head.prev = new_node
            self.head = new_node

        self.size += 1

    def insert(self, val: Any, index: int) -> None:
        """
        Inserts a new node at a given index
        :param val: value of node to insert
        :param index: position at which to insert
        :return: None
        >>> linked_list = DoublyLinkedList()
        >>> linked_list.append(1)
        >>> linked_list.append(2)
        >>> linked_list.append(3)
        >>> linked_list.append(4)
        >>> linked_list.insert(5, 2)
        >>> linked_list.insert(6, 1)
        >>> print(linked_list)
        [ 1 6 2 5 3 4 ]
        """

        if index < 0 or index > self.size:
            raise IndexError("Index out of bounds")
        elif index == 0:
            self.push(val)
        elif index == self.size:
            self.append(val)
        else:
            curr = self.head
            for _ in range(0, index - 1):
                curr = curr.next

            new_node = Node(val)
            new_node.prev = curr
            new_node.next = curr.next

            curr.next = new_node

            self.size += 1

    def remove_tail(self) -> Any:
        """
        Removes the last node
        :return: value of deleted node
        >>> linked_list = DoublyLinkedList()
        >>> linked_list.append(1)
        >>> linked_list.append(2)
        >>> linked_list.remove_last()
        >>> print(linked_list)
        [ 1 ]
        """

        if self.is_empty():
            raise IndexError("Removing from empty list")
        elif self.size == 1:
            retval = self.head.val
            self.head = None
        else:
            curr = self.head
            while curr.next.next is not None:
                curr = curr.next

            retval = curr.next.val
            curr.next = None

        self.size -= 1
        return retval

    def remove_head(self) -> Any:
        """
        Removes the first node
        :return: value of deleted node
        >>> linked_list = DoublyLinkedList()
        >>> linked_list.append(1)
        >>> linked_list.append(2)
        >>> print(linked_list)
        [ 2 ]
        """

        if self.is_empty():
            raise IndexError("Pop from empty list")
        else:
            retval = self.head.val
            new_head = self.head.next

            self.head = None
            self.head = new_head

        self.size -= 1
        return retval

    def delete(self, index: int) -> Any:
        """
        Remove node at a given index
        :param index: position of node to delete
        :return: value of deleted node
        >>> linked_list = DoublyLinkedList()
        >>> linked_list.append(1)
        >>> linked_list.append(2)
        >>> linked_list.append(3)
        >>> linked_list.append(4)
        >>> linked_list.delete(2)
        >>> print(linked_list)
        [ 1 2 4 ]
        """

        if index < 0 or index > self.size:
            raise IndexError("Index out of bounds")
        elif index == 0:
            return self.pop()
        elif index == self.size:
            return self.remove_last()
        else:
            curr = self.head
            for _ in range(0, index - 1):
                curr = curr.next

            retval = curr.next.val

            if curr.next.next is not None:
                curr.next.next.prev = curr

            curr.next = curr.next.next

        self.size -= 1
        return retval

    def get(self, index: int) -> Any:
        """
        Get value of node at given index
        :param index: position of node to find
        :return: value of node at index
        >>> linked_list = DoublyLinkedList()
        >>> linked_list.append(1)
        >>> linked_list.append(2)
        >>> linked_list.append(3)
        >>> print(linked_list.get(1))
        2
        """

        if self.is_empty() or index < 0 or index >= self.size:
            raise IndexError("Index out of bounds")
        else:
            curr = self.head
            for _ in range(0, index):
                curr = curr.next

            return curr.val

    def reverse(self) -> None:
        """
        Reverses the Doubly Linked List
        :return: None
        >>> linked_list = DoublyLinkedList()
        >>> linked_list.append(1)
        >>> linked_list.append(2)
        >>> linked_list.append(3)
        >>> linked_list.append(4)
        >>> linked_list.reverse()
        >>> print(linked_list)
        [ 4 3 2 1 ]
        """

        if not self.size <= 1:
            curr = self.head
            temp = None
            while curr is not None:
                temp = curr.prev
                curr.prev = curr.next
                curr.next = temp

                curr = curr.prev

            self.head = temp.prev


def main():

    # Driver code to construct the following linked list:
    #
    # None <- 1 -> <- 2 -> <- 3 -> <- 4 -> <- 5 -> None

    linked_list = DoublyLinkedList()
    linked_list.append(1)
    linked_list.append(2)
    linked_list.append(3)
    linked_list.append(4)
    linked_list.append(5)

    # Demonstrate basic operations on linked list

    linked_list.push(0)
    linked_list.insert(6, 5)
    linked_list.remove_tail()
    linked_list.remove_head()
    linked_list.delete(4)
    print(linked_list.get(2))

    linked_list.append("A")

    print(linked_list)


    # Reverse method: swaps each prev and next pointer
    # using a temp node to store the prev node

    # None <- old_head -> <- node_1 -> ... <- temp -> <- old_tail -> None

    # On the final step, curr is at the old_tail node, meaning temp is the prev
    # (the second node of the new list, second last of old list)
    # Set the head variable to be the prev value of the temp node (old_tail)

    linked_list.reverse()

    print(linked_list)


if __name__ == "__main__":
    main()
